### PR TITLE
170622: Azure APIM integration

### PR DESCRIPTION
**Pull request by chrissmKainos**

### CODE REVIEW

File: app/api-requests/rpa-api/base.js [hunk 1]
```diff
@@ -1,11 +1,13 @@
 const wreck = require('@hapi/wreck')
 const session = require('../../session')
 const { tokens } = require('../../session/keys')
+const config = require('../../config')
 
 const get = async (hostname, url, request, headers = {}) => {
   const token = session.getToken(request, tokens.accessToken)
-  headers.Authorization = token
-  console.log(`${new Date().toISOString()} Request Message to RPA: ${JSON.stringify(`${hostname}${url}`)}`)
+  headers['X-Forwarded-Authorization'] = token
```
Comment (liamwilson-kainos) [inline app/api-requests/rpa-api/base.js:8]: Could we create a little constants file for APIM headers please?

File: app/api-requests/rpa-api/base.js [hunk 2]
```diff
@@ -1,11 +1,13 @@
...
 const session = require('../../session')
 const { tokens } = require('../../session/keys')
+const config = require('../../config')
 
 const get = async (hostname, url, request, headers = {}) => {
   const token = session.getToken(request, tokens.accessToken)
-  headers.Authorization = token
-  console.log(`${new Date().toISOString()} Request Message to RPA: ${JSON.stringify(`${hostname}${url}`)}`)
+  headers['X-Forwarded-Authorization'] = token
+  headers['Ocp-Apim-Subscription-Key'] = config.authConfig.ruralPaymentsAgency.ocpApimSubscriptionKey
```
Comment (liamwilson-kainos) [inline app/api-requests/rpa-api/base.js:9]: could we put this key into the apim config section?

File: app/config/auth.js [hunk 3]
```diff
@@ -40,7 +48,15 @@ const authConfig = {
     hostname: process.env.RPA_HOST_NAME,
     getPersonSummaryUrl: process.env.RPA_GET_PERSON_SUMMARY_URL,
     getOrganisationPermissionsUrl: process.env.RPA_GET_ORGANISATION_PERMISSIONS_URL,
-    getOrganisationUrl: process.env.RPA_GET_ORGANISATION_URL
+    getOrganisationUrl: process.env.RPA_GET_ORGANISATION_URL,
+    ocpApimSubscriptionKey: process.env.RPA_OCP_APIM_SUBSCRIPTION_KEY
```
Comment (liamwilson-kainos) [inline app/config/auth.js:52]: can we move this one to the apim section?

File: app/routes/auth/signin-oidc.js [hunk 4]
```diff
@@ -46,10 +46,12 @@ module.exports = [{
       try {
         await auth.authenticate(request, session)
 
-        const personSummary = await getPersonSummary(request)
+        const apimAccessToken = await auth.getClientCredentials(request)
+
+        const personSummary = await getPersonSummary(request, apimAccessToken)
```
Comment (liamwilson-kainos) [inline app/routes/auth/signin-oidc.js:51]: something that was on my mind in general - can we add a configurable timeout to the API calls via Wreck please? I think we better do this before we forget - maybe 10seconds until we have a better value?

**Review by liamwilson-kainos — CHANGES_REQUESTED** [review]

File: app/routes/auth/signin-oidc.js (hunk 4 above)
Comment (chrissmKainos) [reply app/routes/auth/signin-oidc.js:51]: Code updated - new config value added with default 10000 (10 secs)

`config.wreckHttp.timeoutMilliseconds`

File: app/api-requests/rpa-api/base.js (hunk 2 above)
Comment (chrissmKainos) [reply app/api-requests/rpa-api/base.js]: Code updated - config value moved into the apim section

File: app/api-requests/rpa-api/base.js (hunk 1 above)
Comment (chrissmKainos) [reply app/api-requests/rpa-api/base.js]: Code updated - constant file added and used by base.js

**Review by liamwilson-kainos — APPROVED** [review]

**Pull request merged**